### ISSUE TITLE
Reset Template Feature

### ISSUE THREAD

**Issue by d-e-v-esh** [accordproject/web-components#315]
# Feature Request 🛍️

Package: Contract Editor
It is a set of three features:
1. Reset Variables - This would reset all the variables to the default that was in the template.
2. Reset Contract - This would discard any changes made to the contract body itself. Doing this would reset the whole body of the contract but leave the updated variables intact.
3. Reset Template - This would reset both the Variables and the Contract Body to the default state like how it was upon import.

## Use Case

This feature is going to be very useful when we mess something up too bad and want a clean start. 
 - Sometimes we might want to do the reset to just the variables. 
 - Sometimes we might need to reset the body of the contract and leave the variables intact as we have filled them up correctly.
 - And sometimes we might just need a hard reset of the entire template. Having this as a feature is going to be very handy so that we don't have to delete and re-import the entire template.

## Detailed Description

This can be implemented as a set of icons on the template around where the `Test`, `Edit`, and `Delete` icons are implemented.
Or this can be a single icon next to those icons which can open up a dropdown on hover which will contain these 3 reset options or at last they can be inside a right-click menu.

**Comment by Michael-Grover** [comment]
@d-e-v-esh Thank you for coming up with the concept for this feature. This feature request was discussed in today's technology working group call, and the participants were not sure about the value this feature would bring to the editor. The consensus was this is not a high priority right now.

We also discussed that consumers of the contract editor can handle the state of the editor, saving the contents of the editor, etc themselves.

Can you summarize the motivation for this feature, or reframe the use cases to help us better understand how this feature would improve the editor?

@dselman @jeromesimeon feel free to weigh in if I missed anything about the discussion.

**Comment by d-e-v-esh** [comment]
@Michael-Grover I think it is just a matter of convenience and efficiency that undo button is just not capable of providing in a lot of situations.

My thought process was that there would be multiple templates in a single contract. The undo button is just going to provide us with a single stack of actions that we performed which we can undo.

Assuming that the undo stack is not trashed out every time the user logs out and logs in the next day if we are making changes in multiple templates in our file simultaneously then to get to an earlier desired position, we might need to undo more than we want which will change the contents of those templates that we don't want to touch anymore. That is why we give the users the option to choose if they want to reset just the variables or the rest of the body of the current template -> providing them more flexibility. With this feature, we can divide each template in a way where resetting one template would not affect the other ones.

There are arrows on the left side of each template which helps us to move them up and down in the contract making them more modular in nature. I think this is the same as that but with the editing function. This feature would be a more convenient way of undo when a particular template was not changed a lot from default and the user want to go back to the default position without thinking too much about it.

**Comment by irmerk** [comment]
This is a very intriguing idea! I really like it. I wasn't on the woking group call yesterday so I'll need to catch the recording. Though, my impression is this would be a very useful feature. I would expect this kind of functionality as an AP user.

**Comment by jeromesimeon** [comment]
I do find it intriguing too. I think some of the questions that were raised included: reset to what state? What happens for long drafting times (which is common on contract) do you need to reset to "previous version" which is something a lot more complicated.

If it's just convenience for developers can it be done inside the storybook rather than in the Contract Editor?

**Comment by d-e-v-esh** [comment]
@irmerk Thank you! Yes, this is not a high-priority feature but I think eventually this would be very necessary for someone who wants to do any serious long contract work on this.

**Comment by d-e-v-esh** [comment]
@jeromesimeon I think initially it can be reset to the initial state (how it looks when we first import it). Later on, **if** we want to develop it further, I think we can develop a feature where we can let the users create a checkpoint for each template that would store how it used to look at a certain point of time, and then the user can jump back and forth in the timeline if they want.

**Comment by Michael-Grover** [comment]
> That is why we give the users the option to choose if they want to reset just the variables or the rest of the body of the current template 

It's my understanding that a user can't change the body of a clause template, they can only change variables, so I'm confused about how a user would reset the body of a template. Since the only things that can be changed are the variables, if they really want to reset those to the defaults, can't they just delete the template and create a new one? Also, is there a need to reset a template to the default variables? Usually the defaults are just placeholders like "Widget, "attachment X", etc. The default variable values are meant to be changed to something else, so I'm not sure what the need to restore the defaults would be.

**Comment by d-e-v-esh** [comment]
@Michael-Grover I assume lawyers and people in the law field would be using AP the most, so they should have the option to build on top of the template or take things off from it according to the needs of their clients. Are we not giving them the option to edit the body?

**Comment by Michael-Grover** [comment]
@d-e-v-esh when you say "template" are you referring to the entire document text, or the text within a Clause Template like "acceptance of delivery"? Currently users can not edit non-variable text of a clause template from within the contract editor. They would have to manipulate the template file directly in a program like VS code. I agree that's not ideal, but it's a much larger issue.
![image](https://user-images.githubusercontent.com/40569793/112493259-63723680-8d58-11eb-9b79-7561bb2988d2.png)

**Comment by d-e-v-esh** [comment]
@Michael-Grover I think the users can currently edit the template in the storybook. There is a `lockText` option and we just need to uncheck the box. It is a little buggy so sometimes it does not work.

![chrome_NUF8tWGLxB](https://user-images.githubusercontent.com/59534570/112494623-3601a900-8da9-11eb-8bbc-f20900387564.png)

**Comment by Michael-Grover** [comment]
@d-e-v-esh I think changing the text breaks the template, which is why there's a red outline around it. The error logger component would show what error the template has.

**Comment by d-e-v-esh** [comment]
@Michael-Grover Yes, I think it does break the template. I don't know enough about the internals of how it works and how it should work so I can't say anything about that. Putting even an extra space gets the redlining going.

**Comment by jeromesimeon** [comment]
> @Michael-Grover Yes, I think it does break the template. I don't know enough about the internals of how it works and how it should work so I can't say anything about that. Putting even an extra space gets the redlining going. I didn't see anything the error logger.

I thought the clause text would be read-only? But maybe there is an option.

And yes this is the intended behaviour: if you change the text you may change the legal meaning so you would want to create a new Clause.

Concretely we parse the text of the clause (with its variable) against the grammar for that template. In your example the one here: https://templates.accordproject.org/acceptance-of-delivery@0.14.1.html

**Comment by K-Kumar-01** [comment]
@jeromesimeon @irmerk @Michael-Grover 
If no one is currently working on this issue, I can have a go at it. I would discuss the steps that can be implemented for every particular sub-feature that we are talking about
@d-e-v-esh Is this fine with you if this issue is assigned to me?

**Comment by Michael-Grover** [comment]
The maintainers have decided that this feature is not a high priority right now, based on my previous feedback in the issue comments, and the discussion in yesterday's technology working group call starting at 33:50 in this recording: https://vimeo.com/529064504. Thank you for the submission @d-e-v-esh . Feel free to use the feedback from the Tech WG call as a starting point for new feature ideas.
